### DISCUSSION TITLE
Does arrow-rs support tracking buffer memory usage?

### DISCUSSION THREAD

**Discussion by jp0317** [apache/arrow-rs#5916] [Q&A] [answered]
Hi, i'm wondering if there's an existing interface to track buffer memory usage within arrow-rs. In C++  there's an [arrow::MemoryPool](https://github.com/apache/arrow/blob/main/cpp/src/arrow/memory_pool.h#L108), e.g., in page decompression it's used for decompressor [buffer](https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L269) while in rust the buffer is from [Vec](https://github.com/apache/arrow-rs/blob/master/parquet/src/file/serialized_reader.rs#L414). Tracking memory usage can be helpful for cases where memory control is desired.

**Comment by tustvold** [comment, marked as answer]
We provide methods to get the allocated size of buffers in arrays and these can be hooked into a memory tracking system. DataFusion has examples of this. We don't provide anything first-party in part because Rust doesn't have stable support for custom allocators which would be the ideal way to provide this functionality.

**Comment by jp0317** [comment]
thanks for the rely! 

>get the allocated size

would this be too late? i.e., it seems better to prevent insane huge allocation by checking the size to be allocated before real allocation.

**Reply by tustvold** [reply]
We support tracking memory usage, fallible allocation could be achieved by overriding the global allocator.

See also https://github.com/apache/arrow-rs/issues/3960 and https://github.com/apache/arrow-rs/issues/6439
